src/sources/kafka.rs

```diff
@@ -113,10 +114,13 @@ fn kafka_source(
 
                 async move {
                     match message {
-                        Err(e) => {
-                            Err(error!(message = "Error reading message from Kafka", error = ?e))
+                        Err(error) => {
+                            emit!(KafkaEventFailed{ error: error.clone() });
+                            Err(error!(message = "Error reading message from Kafka", error = ?error))
```

Comment:
@JeanMertz @lukesteensen Should we merge these? Like you did here: https://github.com/timberio/vector/pull/3151/files#diff-2531c6abda293fda0847fd5f5fa6d474R35

Reply:
The difference here is that `error!` returns an actual error object that gets returned in this location, whereas the link you provided does not return a value (or, more precise, returns a unit value).

The `emit!` macro doesn't return a type right now. We could opt to change that so that we can emit an internal event, and have it return something that can be turned into an error type. Not all internal events represent an error, though.

I do agree that it would be nice if we could clean this up a bit. It's something I wrote down on my list of things I'd like to look into when I get back.

Reply:
Actually. Looking at this again, it looks like the `error!` macro _also_ returns a unit type.

This is confusing. I believe there's even a [Clippy lint](https://rust-lang.github.io/rust-clippy/rust-1.45.0/index.html#unit_arg) to disincentivise writing code like this.

I'll make the changes to remove the `error!` call, as it is indeed unneeded.

Reply:
fixed in 1923f50